Add native ARM64 support for rdtsc

RAttab opened the pull request:
Ran into an issue where while building on docker on a Mac, the rdtsc function would not compile due there not being a non-apple implementation for aarch64.

Did a bit of digging and the `cntvct_el0` register seems to be more or less equivalent to rdtsc while also working in user space. Code has been blatantly copied from the Linux kernel (link provided).  Note that we could probably remove the `__APPLE__` branch if this happens to work well as it would be redundant.

Relevant documentation: https://developer.arm.com/documentation/ddi0595/2021-12/AArch64-Registers/CNTVCT-EL0--Counter-timer-Virtual-Count-register

RAttab committed:
Add native ARM64 support for rdtsc

tokenrove commented:
Thanks!  I don't know how I missed this register when I went to look for this before.  Any chance you could test on macOS?  It would be nice to just have the one branch.

Unfortunately I can't accept this change as is, because the code you've copied is GPL-2, which is incompatible with the Apache license.  In this case, I think rewriting it, basically by replacing the comment with a link to or quote from the aarch64 reference, would be sufficient.

Much appreciated, and sorry for that hassle.

RAttab commented:
Adjusted version to remove references to the GPL-2 code. Also ran a quick test on OSX with the following code:

```
#include "c_src/rdtsc.h"
#include <stdio.h>

extern inline void rdtsc(uint64_t *, uint64_t *);

int main(int argc, char *argv[])
{
    (void) argc, (void) argv;

    uint64_t t = 0, s = 0;
    rdtsc(&t, &s);

    printf("%08llx\n", t);
    return 0;
}
```

Which gave the following results:
```
remi@Remis-MBP ~/code/granderl $ gcc -o rdtsc-test test.c
remi@Remis-MBP ~/code/granderl $ ./rdtsc-test
2af6eebf658f
remi@Remis-MBP ~/code/granderl $ ./rdtsc-test
2af6efee9227
remi@Remis-MBP ~/code/granderl $ ./rdtsc-test
2af6f0e1cea3
remi@Remis-MBP ~/code/granderl $ ./rdtsc-test
2af6f1c75188
remi@Remis-MBP ~/code/granderl $ ./rdtsc-test
2af6f4333399
```

So everything seems to work fine and I removed the `__APPLE__` branch.

Let me know if there's any other tweaks required :)

tokenrove commented:
I mean, your commit message could have a bit more detail :-) but no, that's great, thank you.  Apparently this could trap if some register gets disabled but if it currently works everywhere, we'll wait for someone to complain when some macOS update inevitably breaks things.
